Enhanced append: method with self-check, size update, and link clearing with single Linked List Merging

Tanmay-Tripathi opened the pull request:
Pull Request Description:
This PR improves the append: method in CTLinkedList by ensuring proper merging of two linked lists.

Changes & Improvements:
Handles cases where the current list is empty by adopting aCollection's links.
Properly links the last node of the current list to the first node of aCollection.
Updates the lastLink reference after appending.
Adds clear, concise comments for better readability.
This enhancement ensures smooth appending of linked lists while maintaining integrity.

Commit:
Enhanced append: method with self-check, size update, and link clearing

StephanEggermont commented:
Thanks for the contribution. When looking at the changed method, I notice that the parameter name suggests that every kind of collection can be added 'aCollection', while the sending of #firstLink to that parameter suggests that only linked lists can be used. 
![image](https://github.com/user-attachments/assets/a9c166c5-f67c-4170-8a12-db01ae8815c9)

Tanmay-Tripathi commented:
Ohh thank you for pointing that out i may have overlooked this i will
modify and get back to you shortly.
Meanwhile i wanted to ask about further processes that will help me to work
under your guidance for the gsoc . Like the proposal and the steps i have
to follow to get selected as contributor for this project.

Best wishes,
Tanmay Tripathi

On Wed, Mar 5, 2025, 14:54 Stephan Eggermont ***@***.***>
wrote:

> Thanks for the contribution. When looking at the changed method, I notice
> that the parameter name suggests that every kind of collection can be added
> 'aCollection', while the sending of #firstLink to that parameter suggests
> that only linked lists can be used.
> image.png (view on web)
> <https://github.com/user-attachments/assets/a9c166c5-f67c-4170-8a12-db01ae8815c9>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700337834>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOHZV7Q3VO47ZKCH7I32S27EHAVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQGMZTOOBTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: StephanEggermont]*StephanEggermont* left a comment
> (Ducasse/Containers-LinkedList#8)
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700337834>
>
> Thanks for the contribution. When looking at the changed method, I notice
> that the parameter name suggests that every kind of collection can be added
> 'aCollection', while the sending of #firstLink to that parameter suggests
> that only linked lists can be used.
> image.png (view on web)
> <https://github.com/user-attachments/assets/a9c166c5-f67c-4170-8a12-db01ae8815c9>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700337834>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOHZV7Q3VO47ZKCH7I32S27EHAVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQGMZTOOBTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

StephanEggermont commented:
Further, it was not obviously clear to me if the intended behavior was supposed to be destructive to the parameter collection, reusing the link objects. A test demonstrating that, and its pitfalls/mitigation (copy), would be good.

Tanmay-Tripathi commented:
Ok i will make everything much more clear and effective and get back to you
by tonight
I have understood your feedbacks and i will work through it

On Wed, Mar 5, 2025, 15:07 Stephan Eggermont ***@***.***>
wrote:

> Further, it was not obviously clear to me if the intended behavior was
> supposed to be destructive to the parameter collection, reusing the link
> objects. A test demonstrating that, and its pitfalls/mitigation (copy),
> would be good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700371394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOFG7JNW4GWT6Y5SGLL2S3AWHAVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQGM3TCMZZGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: StephanEggermont]*StephanEggermont* left a comment
> (Ducasse/Containers-LinkedList#8)
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700371394>
>
> Further, it was not obviously clear to me if the intended behavior was
> supposed to be destructive to the parameter collection, reusing the link
> objects. A test demonstrating that, and its pitfalls/mitigation (copy),
> would be good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2700371394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOFG7JNW4GWT6Y5SGLL2S3AWHAVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBQGM3TCMZZGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Tanmay-Tripathi commented:
Hello @StephanEggermont @Ducasse 
Should we rename aCollection to aLinkedList for clarity? Also, regarding the destructive behavior, do you prefer copying aLinkedList before appending, documenting the current behavior, or adding a flag for optional copying? Let me know your thoughts, and I’ll update the PR accordingly.

Ducasse commented:
Did you validate the changes by adding new tests?

Tanmay-Tripathi commented:
Basically i did it locally not on the test file, i am on it will get back
with the updated test codes or methods along with the point stated.

May i know will this project be given priority as i am looking forward to
this with you

On Thu, Mar 6, 2025, 01:05 StéphaneDucasse ***@***.***> wrote:

> Did you validate the changes by adding new tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOBG7H4GVY5HG33IBZL2S5GZ7AVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRHA4DONRZGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: Ducasse]*Ducasse* left a comment (Ducasse/Containers-LinkedList#8)
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>
>
> Did you validate the changes by adding new tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOBG7H4GVY5HG33IBZL2S5GZ7AVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRHA4DONRZGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

StephanEggermont commented:
Renaming the parameter sounds like a good idea. I do not use linked lists enough to have a clear preference for destructive or non-destructive behaviour, but I do have a clear preference for documenting the chosen behaviour in a unit test.

Tanmay-Tripathi commented:
sure i am on it i will be submitting another pr by tomorrow

Commit:
modified the append linked list with renaming aCollection to aLinkedList and added few more test cases

Tanmay-Tripathi commented:
Hello there @StephanEggermont @Ducasse sorry i was busy with the office work but i have figured out a logic which i have tested too in the new commit 

- Renamed `aCollection` to `aLinkedList` for clarity.  
- Implemented a safe approach by copying `aLinkedList` before appending to avoid modifying the original list.  
- Added a check to prevent self-appending to avoid infinite loops. (basically a test case funtion)  
- Updated comments for better readability.  
- Added tests to verify:  
  - Appending two linked lists with variable size 
  - Self-appending prevention base case
  - Ensuring the original list remains unchanged after append.  
  
  Please have a look and tell me if you think this is the correct approach? I am looking forward for this project.

Tanmay-Tripathi commented:
Hello Stephane,
I have added a new commit in the pull request please have a look it
contains

   - Renamed aCollection to aLinkedList for clarity.
   - Implemented a *safe approach* by copying aLinkedList before appending
   to avoid modifying the original list.
   - Added a check to *prevent self-appending* to avoid infinite loops.
   - Updated comments for better readability.
   - Added tests to verify:
      - Appending two linked lists of different size
      - Self-appending prevention.
      - Ensuring the original list remains unchanged after append.


Please have a look at it and give me your valuable feedback.

Best
Tanmay Tripathi

On Thu, Mar 6, 2025 at 1:05 AM StéphaneDucasse ***@***.***>
wrote:

> Did you validate the changes by adding new tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOBG7H4GVY5HG33IBZL2S5GZ7AVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRHA4DONRZGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: Ducasse]*Ducasse* left a comment (Ducasse/Containers-LinkedList#8)
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>
>
> Did you validate the changes by adding new tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ducasse/Containers-LinkedList/pull/8#issuecomment-2701887696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2MLGOBG7H4GVY5HG33IBZL2S5GZ7AVCNFSM6AAAAABYJV5DF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRHA4DONRZGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jordanmontt commented on src/Containers-LinkedList/CTLinkedList.class.st:
```diff
@@ -184,15 +184,35 @@ CTLinkedList >> addLast: aLinkOrObject [
+    aLinkedList isEmpty ifTrue: [ ^ self ].
+
+    "Prevent self-appending."
+    aLinkedList == self ifTrue: [ ^ self error: 'Cannot append list to itself' ].
```
I'm not sure about this. Why one could not add the list (duplicate it)?

Tanmay-Tripathi replied:
so i was thinking that if i have a list 1->3->4 then what is the point of appending the same list in one that is just duplicacy of the data

jordanmontt replied:
Each time that someone add an element to the collection we will for the price for the checking. Why not allowing it? the other collection available in Pharo accept this behavior.

For example look:

```st
| col |
col := OrderedCollection withAll: #(1 2 3).
col addAll: col.
col "an OrderedCollection(1 2 3 1 2 3)"
```

Tanmay-Tripathi replied:
Yes i understood your point and i will get back to it shortly with some improvement and better approach

Tanmay-Tripathi replied:
so for both the suggestions and comments what i can deduce is that 
1. Allow self-appending with element duplication
2. Since firstLink and lastLink are only references, they remain the same whether we copy the list or not. 
so Remove the copy step ?

am i correct @jordanmontt

jordanmontt replied:
Yes, According to me it's correct

Tanmay-Tripathi replied:
ok thank you i will shortly push a better commit as per the above consideration
please have a look at your mail i have sent you a proposal related to the data structures gsoc project

Tanmay-Tripathi replied:
please have a look at the pr i have made the required changes as per the comments

jordanmontt commented on src/Containers-LinkedList/CTLinkedList.class.st:
```diff
@@ -184,15 +184,35 @@ CTLinkedList >> addLast: aLinkOrObject [
+
+    "Copy to avoid modifying the original."
+    | copiedList |
+    copiedList := aLinkedList copy.
```
actually, this is not necessary. Because the copy will only copy the list and not the elements. And the implementation of firstLink only returns the head and lastLink only returns the tail. So I'd say to just not copy

Tanmay-Tripathi replied:
ok i get your point let me try something else to get back with better thing

Tanmay-Tripathi replied:
so according to you what can be the ideal thing to do ?

jordanmontt replied:
to not copy the list

Tanmay-Tripathi committed:
Update CTLinkedList.class.st with removing the copy and allowing the duplicate of list

Tanmay-Tripathi committed:
Update CTLinkedListTest.class.st with the test cases to match the changes in class
